models/userModel.js

```diff
@@ -0,0 +1,236 @@
+    };
+
+    return jwt.sign(payload, SECRET_KEY, JWT_OPTIONS);
+  }
+
+  /** Generate and stores a token.
+   *
+   * Returns a token
+   *
+   * Throws BadRequestError on duplicate emails. **/
+
+  static async createRegistrationToken(userId) {
+    /** Generate 16 random bytes using the OS's CSPRNG*/
+    const randomBytes = crypto.randomBytes(16);
+
+    const plainTextToken = base32Encode(randomBytes, "RFC4648", {
+      padding: false,
+    });
+
+    const expirationTimestamp = new Date(
+      Date.now() + REGISTRATION_TTL * 60 * 1000
+    );
+
+    const hashedToken = await bcrypt.hash(plainTextToken, BCRYPT_WORK_FACTOR);
+
+    await pool.query(
+      `INSERT INTO tokens_registration
+             (hashed_token, fk_user_id, expiration_timestamp)
+             VALUES ($1, $2, $3)
+             `,
+      [hashedToken, userId, expirationTimestamp]
+    );
+    return plainTextToken;
+  }
+
+  /** Validates a registration token.
+   *
+   * Returns a boolean (true/false). **/
+
+  static async validToken({ token, id: userId }) {
+    const result = await pool.query(
+      `SELECT * FROM tokens_registration
+      WHERE fk_user_id = $1`,
+      [userId]
+    );
+
+    if (!result.rows[0]) {
+      throw new BadRequestError(
+        `Your registration link does not exist. Ensure the original link we e-mailed you has not been modified.`
+      );
+    }
+
+    const isValid = await bcrypt.compare(token, result.rows[0].hashed_token);
+
+    if (!isValid) {
+      throw new BadRequestError(
+        `Your registration link does not exist. Ensure the original link we e-mailed you has not been modified.`
+      );
+    }
+
+    if (Date.now() > result.rows[0].expiration_timestamp) {
+      /** Delete expired token */
+      await pool.query(
+        `DELETE FROM tokens_registration
+        WHERE fk_user_id = $1`,
+        [userId]
+      );
+
+      /** Generate and store new registration token  */
+      const newToken = await User.createRegistrationToken(userId);
+
+      /** Email new registration token  */
+      const userEmail = await pool.query(
+        `SELECT email FROM users
+        WHERE id = $1`,
+        [userId]
+      );
+
+      await sendEmailRegistration({
+        id: Number(userId),
+        email: userEmail.rows[0].email,
+        plainTextToken: newToken,
+      });
+
+      //TODO Replace logic below once I create a React button that user has to click to send new token.
+      throw new BadRequestError(
+        `Your registration link is expired. We've emailed you a new registration token.`
+      );
+      // throw new BadRequestError(
+      //   `Your registration link is expired. If you wish to get a new link, click on 'Send new registration link' button below.`
+      // );
+    }
+    return true;
+  }
+}
+
+//PEER Lawrence, is it best to move the token-related methods above (createAuthToken, createRegistrationToken, and validToken) to a separate Token class or are they fine here?
```

Comment:
If the `token` has an entity, then separate the `token model` from the `user model`